engine/src/main/scala/cromwell/jobstore/JobStoreWriterActor.scala

```diff
@@ -1,96 +1,89 @@
+      val completions = data.toVector.collect({ case CommandAndReplyTo(c, _) => c.completion })
+
+      if (completions.nonEmpty) {
+        val workflowCompletions = completions collect { case w: WorkflowCompletion => w }
+        val completedWorkflowIds = workflowCompletions map { _.workflowId } toSet
+        // Filter job completions that also have a corresponding workflow completion; these would just be
+        // immediately deleted anyway.
+        val jobCompletions = completions.toList collect { case j: JobCompletion if !completedWorkflowIds.contains(j.key.workflowId) => j }
+
+        jsd.writeToDatabase(workflowCompletions, jobCompletions, dbBatchSize) onComplete {
+          case Success(_) =>
+            data map { case CommandAndReplyTo(c, r) => r ! JobStoreWriteSuccess(c) }
+            self ! DbWriteComplete
+          case Failure(regerts) =>
+            log.error("Failed to properly job store entries to database", regerts)
+            data map { case CommandAndReplyTo(_, r) => r ! JobStoreWriteFailure(regerts) }
+            self ! DbWriteComplete
+        }
+      }
+      stay using NoData
+    case Event(DbWriteComplete, curData) =>
+      log.debug("Flush of job store commands complete")
+      goto(WaitingToWrite)
   }
 
   onTransition {
-    case (oldState, newState) =>
-      log.debug(s"Transitioning from $oldState to $newState")
+    case WaitingToWrite -> WritingToDb => self ! FlushBatchToDb
   }
+}
 
-  def writeNextOperationToDatabase(data: JobStoreWriterData): JobStoreWriterData = {
+object JobStoreWriterActor {
 
-    val newData = data.rolledOver
+  case class CommandAndReplyTo(command: JobStoreWriterCommand, replyTo: ActorRef)
 
-    val workflowCompletions = newData.currentOperation collect {
-      case (_, RegisterWorkflowCompleted(wfid)) =>  wfid
-    }
-
-    val jobCompletions = newData.currentOperation collect {
-      case (_, RegisterJobCompleted(jobStoreKey, jobResult)) if !workflowCompletions.contains(jobStoreKey.workflowId) => (jobStoreKey, jobResult)
-    }
+  def props(jobStoreDatabase: JobStore, dbBatchSize: Int, dbFlushRate: FiniteDuration): Props = Props(new JobStoreWriterActor(jobStoreDatabase, dbBatchSize, dbFlushRate)).withDispatcher(EngineDispatcher)
 
-    if (!(workflowCompletions.isEmpty && jobCompletions.isEmpty)) {
-      jsd.writeToDatabase(jobCompletions.toMap, workflowCompletions) onComplete {
-        case Success(_) =>
-          newData.currentOperation foreach { case (actor, message) =>
-            val msg = JobStoreWriteSuccess(message)
-            actor ! msg
-          }
-          self ! WriteComplete
-        case Failure(reason) =>
-          log.error(s"Failed to write to database: $reason")
-          newData.currentOperation foreach { case (actor, message) => actor ! JobStoreWriteFailure(reason) }
-          self ! WriteComplete
-      }
+  implicit class FancyJobStoreWriterCommand(val command: JobStoreWriterCommand) extends AnyVal {
```

Comment:
I wrote it up and it's a little awkward as there's an abstract member in the trait and then overrides as the only functions in the concrete types.